### DISCUSSION TITLE
An Option for Better DirectPlay Audio Codec Selection

### DISCUSSION THREAD

**Discussion by wwall3r** [jellyfin-archive/jellyfin-roku-legacy#1382] [Ideas]
My setup is a bit unorthodox: I do not allow transcoding. I've got Unraid running on a 12 year old Celeron that just won't die and it has no GPU to speak of either. Any transcoding is done up front, once, and forever afterward uses DirectPlay.

I recently bought a Roku Ultra for its nice codec support, wifi, and overall snappiness compared to the old box I had. The other nice feature is that it supports wireless headphones. I was previously laying a 12ft headphone extension cable across the room to my receiver when I wanted to catch a movie late at night.

Difficulty: the Roku only supports DTS via HDMI passthrough. Which is great when connected to the receiver, but won't play at all when wireless headphones are connected. Most of my mobile devices will decode DTS locally and send it over to the headphones, so I imagine this is more about Roku wanting to skip out on Dolby licensing fees?

Anyway, this means that all my files with only DTS audio need another audio track with plain AAC. That's doable. I already wrote a script to do that. The poor Celeron is going to have its cores full for a while.

The issue here is that after having connected the headphones and pulling up a video, Jellyfin always selects the first audio track, regardless of whether that track is actually supported by the current hardware configuration. That might be fine when transcoding is available, but when it isn't, it makes the user go through the options to find one themselves (and it also doesn't save that choice, which is arguably fine).

I already did some poking around and got a dev build up and running which selects the first track which the device reports it can decode, but it really ought to be based on the user's transcoding permissions. So: `if only direct play, only pick audio streams which are supported`. 

Just throwing this out there to see if you all would be amenable to a feature like that before I put any more work into it.

**Comment by cewert** [comment]
I was about to type up a response telling you what I thought would work but it was so easy that I just made a quick PR to show you what I was thinking :) I haven't tested it yet but [here](https://github.com/jellyfin/jellyfin-roku/pull/1383) it is. Let me know if that's similar to what you were thinking.

So the dev/unstable build fixed your problem with the headphones not working?

**Reply by wwall3r** [reply]
Thanks for the quick response! The local change I made was just a quick and dirty 
```diff
diff --git a/components/movies/MovieDetails.brs b/components/movies/MovieDetails.brs
index 913ccd7..f6ef232 100644
--- a/components/movies/MovieDetails.brs
+++ b/components/movies/MovieDetails.brs
@@ -224,8 +224,10 @@ end sub


 sub SetDefaultAudioTrack(itemData)
+    di = CreateObject("roDeviceInfo")
+
     for i = 0 to itemData.mediaStreams.Count() - 1
-        if itemData.mediaStreams[i].Type = "Audio"
+        if itemData.mediaStreams[i].Type = "Audio" and di.CanDecodeAudio({ Codec: itemData.mediaStreams[i].codec }).result
             m.top.selectedAudioStreamIndex = i
             setFieldText("audio_codec", tr("Audio") + ": " + itemData.mediaStreams[i].displayTitle)
             exit for
```
I will test your PR later tonight to see if it resolves my issues.
